Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -1042,6 +1042,7 @@ public static void Inline(Program program)
       }
       catch (AggregateException ae)
       {
+        ThrowIfCanceled(ae);
```

Comment:
This could probably be handled more cleanly inside the `ae.Flatten().Handle` call below.

Reply:
I don't think so. Calling code will rely on this OperationCanceledException to be thrown instead of an all-purpose AggregateException
Handle will throw the first exception that wasn't handled. Here we don't care about other exceptions if one is an OeprationCanceledException.
Moreover, the existing code to which you point just "cancels out" the exception "OperationCanceledException" because it returns true there, meaning it pretends it was handled. I'm not sure why someone wrote that though.

Reply:
Ah, right, and you want `OperatingCanceledException` to take priority over anything else. (Otherwise you could just change that `true` to a `false`, I think?). That makes sense.

Reply:
Here are some places that catch this exception, and they don't catch `AggregatedException`:

https://github.com/dafny-lang/dafny/blob/master/Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs#L98
https://github.com/dafny-lang/dafny/blob/cf0427b87715fd358cbbffed21d51defbeb58b17/Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs#L86
https://github.com/dafny-lang/dafny/blob/cf0427b87715fd358cbbffed21d51defbeb58b17/Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs#L159

Reply:
Is it is possible for a Task to throw anything other than an `AggregateException`? ([documentation](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library)) It seems to me like those places are incorrectly handling something other than `AggregateException`. I've placed breakpoints in the handlers and ran the entire test-suite but didn't hit the breakpoints.

Also, sometimes rethrowing an `AggregateException` through a call to `ae.Flatten().Handle` and sometimes throwing a `OperationCanceledException` seems like an API problem to me, since then callers of `InferAndVerify` have to handle both exceptions regularly and again if they are nested inside an `AggregateException`.

Reply:
Wow this is fascinating. It looks like you've uncovered a latent issue there.

> Is it is possible for a Task to throw anything other than an AggregateException? I think it's not.

I think it is, quote from https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming:
>Every Task will store a list of exceptions. When you await a Task, the first exception is re-thrown, so you can catch the specific exception type (such as InvalidOperationException). However, when you synchronously block on a Task using Task.Wait or Task.Result, all of the exceptions are wrapped in an AggregateException and thrown. Refer again to Figure 4. The try/catch in MainAsync will catch a specific exception type, but if you put the try/catch in Main, then it will always catch an AggregateException. Error handling is much easier to deal with when you don’t have an AggregateException, so I put the “global” try/catch in MainAsync.

Reply:
Ah, thanks for the info :-) 

Do you have a comment on this?
> sometimes rethrowing an AggregateException through a call to ae.Flatten().Handle and sometimes throwing a OperationCanceledException seems like an API problem to me, since then callers of InferAndVerify have to handle each exception type both as a type level exception and as one nested inside an AggregateException. I think the API for any method should be to only throw AggregateException or to never throw it.

Reply:
How do you feel about the second paragraph in my previous comment?

Reply:
I wanted to make sure I found relevant information to you before commenting on your second paragraph, because I was unsure about the first one.
>Also, sometimes rethrowing an AggregateException through a call to ae.Flatten().Handle and sometimes throwing a OperationCanceledException seems like an API problem to me, since then callers of InferAndVerify have to handle each exception type both as a type level exception and as one nested inside an `AggregateException`.

I agree on that, it would be better to have only one type of exception for an API.
> I think the API for any method should be to only throw `AggregateException` or to never throw it.

I think I agree for now only on the cases when we do rethrow an `AggregateException` like above. Otherwise, calling threads should be able to throw the exceptions they want after receiving an `AggregateException`, they will still be gathered into an `AggregateException` depending on how they had been called.
I changed the code to remove entirely `ThrowIfCanceled` and reimplemented the same flow by just removing the line doing the test on `OperationCanceledException` as @atomb  suggested.
The Dafny test pass, however when debugging, it still pauses execution and displays the `AggregateException` as a "user-unhandled exception", which it did not do with the custom `ThrowIfCanceled`. I'm not sure whether it's a smell or it's just an IDE opinionated about when to stop the debugger or not on exception.